fastestimator/pipeline.py

```diff
@@ -13,442 +13,132 @@
-        self.set_epoch_and_mode(epoch=0, mode="train")
-
-    def set_epoch_and_mode(self, epoch: int, mode: str):
-        self.dataset.set_epoch_and_mode(epoch, mode)
-
-
-def get_signature_epochs(val: Union[Scheduler, Iterable[Any]]) -> List[int]:
-    signature_epochs = {0}
-    if isinstance(val, Scheduler):
-        signature_epochs.update(val.epoch_dict.keys())
-    else:
-        for elem in val:
-            if isinstance(elem, Scheduler):
-                signature_epochs.update(elem.epoch_dict.keys())
-    return sorted(signature_epochs)
 
+class Pipeline:
+    def __init__(self,
+                 train_data: Union[Dataset, DataLoader, Scheduler, tf.data.Dataset, None] = None,
+                 eval_data: Union[Dataset, DataLoader, tf.data.Dataset, None] = None,
+                 test_data: Union[Dataset, DataLoader, tf.data.Dataset, None] = None,
+                 batch_size: Union[int, Scheduler, None] = None,
+                 ops: Union[None, NumpyOp, Scheduler, List[Union[NumpyOp, Scheduler]]] = None,
+                 num_process: Union[int, None] = None):
+        self.data = {x: y for (x, y) in zip(["train", "eval", "test"], [train_data, eval_data, test_data]) if y}
+        self.batch_size = batch_size
+        self.ops = ops
+        self.num_process = num_process
+        self.fe_dataset_exist = False
+        self._verify_inputs()
+        self.ops = [] if ops is None else to_list(ops)
+        self.num_process = os.cpu_count() if num_process is None else num_process
 
-def get_per_epoch(ops: Iterable[Union[T, Scheduler[T]]], epoch: int) -> List[T]:
-    per_epoch = []
-    for op in ops:
-        if isinstance(op, Scheduler):
-            per_epoch.append(op.get_current_value(epoch))
+    def _verify_inputs(self):
+        for mode, dataset in self.data.items():
+            if isinstance(dataset, (RepeatScheduler, EpochScheduler)):
+                for ds in dataset.get_items():
+                    self._verify_dataset(mode, ds)
+            else:
+                self._verify_dataset(mode, dataset)
+        if not self.fe_dataset_exist:
+            assert not self.batch_size, "only support batch_size with built-in dataset in Pipeline"
+            assert not self.ops, "only support ops with built-in dataset in Pipeline"
+            assert not self.num_process, "only support num_process with built-in dataset in Pipeline"
+
+    def _verify_dataset(self, mode, dataset):
+        if isinstance(dataset, Dataset):
+            self.fe_dataset_exist = True
+            #batch_size check
+            assert isinstance(self.batch_size, (RepeatScheduler, EpochScheduler, int)), "unsupported batch_size format"
+            if isinstance(self.batch_size, Scheduler):
+                for batch_size in self.batch_size.get_items():
+                    assert isinstance(batch_size, int), "unsupported batch_size format"
+            #ops check
+            assert isinstance(self.ops, list), "unsupported ops format"
```

Comment:
we're not auto-wrapping single ops into list anymore? (wouldn't have been a problem if self.ops was assigned properly before _verify_inputs)

Reply:
you are right, the assertion is not necessary, removed it.